Core/GDCore/IDE/Events/ExpressionValidator.h

```diff
@@ -56,52 +71,238 @@ class GD_CORE_API ExpressionValidator : public ExpressionParser2NodeWorker {
     node.expression->Visit(*this);
-    if (node.child) node.child->Visit(*this);
+    parentType = currentParentType;
+
+    if (node.child) {
+      node.child->Visit(*this);
+    }
   }
   void OnVisitIdentifierNode(IdentifierNode& node) override {
     ReportAnyError(node);
+    if (parentType == Type::String) {
+      RaiseTypeError(_("You must wrap your text inside double quotes "
+                            "(example: \"Hello world\")."),
+                          node.location);
+    }
+    else if (parentType == Type::Number) {
+      RaiseTypeError(
+          _("You must enter a number."), node.location);
+    }
+    else if (parentType == Type::NumberOrString) {
+      RaiseTypeError(
+          _("You must enter a number or a text, wrapped inside double quotes "
+            "(example: \"Hello world\")."),
+          node.location);
+    }
+    else if (parentType != Type::Object && parentType != Type::Variable) {
+      RaiseTypeError(
+          _("You've entered a name, but this type was expected:") + " " + TypeToSting(parentType),
+          node.location);
+    }
+    childType = parentType;
   }
   void OnVisitObjectFunctionNameNode(ObjectFunctionNameNode& node) override {
     ReportAnyError(node);
   }
   void OnVisitFunctionCallNode(FunctionCallNode& node) override {
+    childType = ValidateFunction(node);
+  }
+  void OnVisitEmptyNode(EmptyNode& node) override {
     ReportAnyError(node);
-    for (auto& parameter : node.parameters) {
-      parameter->Visit(*this);
+    gd::String message;
+    if (parentType == Type::Number) {
+      message = _("You must enter a number or a valid expression call.");
+    } else if (parentType == Type::String) {
+      message = _(
+          "You must enter a text (between quotes) or a valid expression call.");
+    } else if (parentType == Type::Variable) {
+      message = _("You must enter a variable name.");
+    } else if (parentType == Type::Object) {
+      message = _("You must enter a valid object name.");
+    } else {
+      message = _("You must enter a valid expression.");
     }
+    RaiseTypeError(message, node.location);
+    childType = Type::Empty;
   }
-  void OnVisitEmptyNode(EmptyNode& node) override { ReportAnyError(node); }
 
  private:
-  void ReportAnyError(ExpressionNode& node) {
+  enum Type {Unknown = 0, Number, String, NumberOrString, Variable, Object, Empty};
+  Type ValidateFunction(const gd::FunctionCallNode& function);
+
+  void ReportAnyError(const ExpressionNode& node) {
     if (node.diagnostic && node.diagnostic->IsError()) {
       errors.push_back(node.diagnostic.get());
     }
   }
 
+  void RaiseError(const gd::String &type, 
+      const gd::String &message, const ExpressionParserLocation &location) {
+    auto diagnostic = std::move(gd::make_unique<ExpressionParserError>(
+        type, message, location));
+    errors.push_back(diagnostic.get());
+    supplementalErrors.push_back(std::move(diagnostic));
```

Comment:
It's to keep the errors alive, but it probably doesn't work as leaving the helper method could destroy the validator.

Reply:
Using static constants for error messages will probably make the code hard to read, but I don't see any other solution.

Reply:
Could you store a copy of the diagnostics?

In any case I would really really avoid this pattern of storing raw pointers in a vector and the unique_ptr in a supplementalErrors.
This makes the intention not clear (one could drop the supplementalError, because they think it's "just extra stuff I don't care about"), invalidating all pointers stored in the validator.
We probably need another approach.

Reply:
Could you store a copy of the diagnostics?

In any case I would really really avoid this pattern of storing raw pointers in a vector and the unique_ptr in a supplementalErrors.
This makes the intention not clear (one could drop the supplementalError, because they think it's "just extra stuff I don't care about"), invalidating all pointers stored in the validator.
We probably need another approach.

Reply:
> Could you store a copy of the diagnostics?

Do you mean to copy errors from the AST to return a vector of unique pointers?

> one could drop the supplementalError, because they think it's "just extra stuff I don't care about"

By "one", do you mean maintainers of this class? because `supplementalErrors`is a protected member (and should probably be private).